src/session.ts

```diff
@@ -232,12 +245,29 @@ export class Session {
             if (allowModify) {
                 request = response.request.clone()
             }
-            // If signatures were returned, append them
-            if (response.signatures) {
+
+            // If signatures were returned, record them in the response.
+            if (response.signatures?.length) {
+                // Merge new signatures alongside existing signatures into the TransactResult.
                 result.signatures = [...result.signatures, ...response.signatures]
+
+                // Recover the keys used to generate the signatures at the time of the request.
+                const recoveredKeys = response.signatures.map((signature) => {
+                    const requestTransaction = request.getRawTransaction()
+                    const requestDigest = requestTransaction.signingDigest(this.chain.id)
+                    return signature.recoverDigest(requestDigest)
+                })
+
+                // Merge newly discovered keys into the TransactResult.
+                result.keys = [...result.keys, ...recoveredKeys]
```

Comment:
When signatures are returned by `beforeSign` hooks, recover the public key from the signature using the current digest of the transaction.